File: src/main/cpp/console_printer.cpp
// Copyright (c) 2023-2023. NESP Technology.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License. You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed
// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License
// for the specific language governing permissions and limitations under the License.

//
// Created by jinzhaolu on 2023/2/11.
//
#include <iostream>
#include "../include/nesp_logger.h"
#include "../include/console_printer.h"

using namespace NespLogger;

//the following are UBUNTU/LINUX ONLY terminal color codes.
#define RESET   "\033[0m\033[0m"
#define BLACK   "\033[0m\033[30m"      /* Black */
#define RED     "\033[0m\033[31m"      /* Red */
#define GREEN   "\033[0m\033[32m"      /* Green */
#define YELLOW  "\033[0m\033[33m"      /* Yellow */
#define BLUE    "\033[0m\033[34m"      /* Blue */
#define MAGENTA "\033[0m\033[35m"      /* Magenta */
#define CYAN    "\033[0m\033[36m"      /* Cyan */
#define WHITE   "\033[0m\033[37m"      /* White */
#define BOLDBLACK   "\033[1m\033[30m"      /* Bold Black */
#define BOLDRED     "\033[1m\033[31m"      /* Bold Red */
#define BOLDGREEN   "\033[1m\033[32m"      /* Bold Green */
#define BOLDYELLOW  "\033[1m\033[33m"      /* Bold Yellow */
#define BOLDBLUE    "\033[1m\033[34m"      /* Bold Blue */
#define BOLDMAGENTA "\033[1m\033[35m"      /* Bold Magenta */
#define BOLDCYAN    "\033[1m\033[36m"      /* Bold Cyan */
#define BOLDWHITE   "\033[1m\033[37m"      /* Bold White */

void ConsolePrinter::print(Logger::LogRecord logRecord) {
    const Logger::LogRecord record = logRecord;
    string color;
    if (record.level == Logger::Level::INFO || record.level == Logger::Level::DEBUG) {
        color = RESET;
    } else if (record.level == Logger::Level::WARN) {
        color = YELLOW;
    } else if (record.level == Logger::Level::ERROR) {
        color = RED;
    } else if (record.level == Logger::Level::FATAL) {
        color = BOLDRED;
    }
    cout << color << record.messageFormatted << endl;
}

ConsolePrinter::~ConsolePrinter() = default;

File: rust/src/printer/console_printer.rs
/*
 * Copyright (c) 2023. NESP Technology.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License
 * for the specific language governing permissions and limitations under the License.
 *
 */

use crate::level;
use crate::level::Level;
use crate::log_record::LogRecord;
use crate::printer::printer::Printer;

const RESET: &'static str = "\033[0m\033[0m";
const BLACK: &'static str = "\033[0m\033[30m"; /* Black */
const RED: &'static str = "\033[0m\033[31m"; /* Red */
const GREEN: &'static str = "\033[0m\033[32m"; /* Green */
const YELLOW: &'static str = "\033[0m\033[33m"; /* Yellow */
const BLUE: &'static str = "\033[0m\033[34m"; /* Blue */
const MAGENTA: &'static str = "\033[0m\033[35m"; /* Magenta */
const CYAN: &'static str = "\033[0m\033[36m"; /* Cyan */
const WHITE: &'static str = "\033[0m\033[37m"; /* White */
const BOLD_BLACK: &'static str = "\033[1m\033[30m"; /* Bold Black */
const BOLD_RED: &'static str = "\033[1m\033[31m"; /* Bold Red */
const BOLD_GREEN: &'static str = "\033[1m\033[32m"; /* Bold Green */
const BOLD_YELLOW: &'static str = "\033[1m\033[33m"; /* Bold Yellow */
const BOLD_BLUE: &'static str = "\033[1m\033[34m"; /* Bold Blue */
const BOLD_MAGENTA: &'static str = "\033[1m\033[35m"; /* Bold Magenta */
const BOLD_CYAN: &'static str = "\033[1m\033[36m"; /* Bold Cyan */
const BOLD_WHITE: &'static str = "\033[1m\033[37m"; /* Bold White */

pub struct ConsolePrinter;

impl ConsolePrinter {
    pub fn new() -> ConsolePrinter {
        ConsolePrinter {}
    }
}

impl Printer for ConsolePrinter {
    fn print(&self, record: LogRecord) {
        let level = record.level;
        let mut color: &str = RESET;

        match level {
            Level::Debug => {
                color = RESET;
            }
            Level::Info => {
                color = RESET;
            }
            Level::Warn => {
                color = YELLOW;
            }
            Level::Error => {
                color = RED;
            }
            Level::Fatal => {
                color = BOLD_RED;
            }
        }

        println!("{}", color.to_string() + &*record.message_formatted);
    }
}
